File: comfyui-extension/comfyui-edition-alt/ui/src/schemas/settings/recent-projects-schema.ts
/**
 * Recent Projects Schema
 *
 * Zod schemas for validating recent projects list stored in localStorage.
 * Includes comprehensive validation constraints for security and data integrity.
 */

import { z } from "zod";
import {
  entityId,
  MAX_NAME_LENGTH,
} from "../shared-validation";

// ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
// Primitives
// ════════════════════════════════════════════════════════════════════════════

const nonNegativeInt = z.number().int().nonnegative();

// ════════════════════════════════════════════════════════════════════════════
// Recent Project Schema
// ════════════════════════════════════════════════════════════════════════════

export const RecentProjectSchema = z.object({
  id: entityId,
  name: z.string().min(1).max(MAX_NAME_LENGTH).trim(),
  thumbnail: z.string().max(10 * 1024 * 1024).optional(), // Max 10MB base64 thumbnail
  modifiedAt: nonNegativeInt.max(2147483647000), // Max timestamp (year 2038)
}).strict();

export type RecentProject = z.infer<typeof RecentProjectSchema>;

// ════════════════════════════════════════════════════════════════════════════
// Recent Projects Array Schema
// ════════════════════════════════════════════════════════════════════════════

export const RecentProjectsSchema = z.array(RecentProjectSchema).max(100); // Max 100 recent projects

export type RecentProjects = z.infer<typeof RecentProjectsSchema>;

// ════════════════════════════════════════════════════════════════════════════
// Validation Helpers
// ════════════════════════════════════════════════════════════════════════════

export function validateRecentProject(data: unknown): RecentProject {
  return RecentProjectSchema.parse(data);
}

export function safeValidateRecentProject(data: unknown) {
  return RecentProjectSchema.safeParse(data);
}

export function validateRecentProjects(data: unknown): RecentProjects {
  return RecentProjectsSchema.parse(data);
}

export function safeValidateRecentProjects(data: unknown) {
  return RecentProjectsSchema.safeParse(data);
}

File: comfyui-extension/leancomfy/haskell/Lattice/Schemas/Settings/RecentProjectsSchema.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : Lattice.Schemas.Settings.RecentProjectsSchema
Description : Recent projects validation
Copyright   : (c) Lattice, 2026
License     : MIT

Validates recent projects list stored in localStorage.

Source: leancomfy/lean/Lattice/Schemas/Settings/RecentProjectsSchema.lean
-}

module Lattice.Schemas.Settings.RecentProjectsSchema
  ( -- * Constants
    maxRecentProjects
  , maxThumbnailSize
  , maxTimestamp
    -- * Recent Project
  , RecentProject(..)
  , validateRecentProject
  , safeValidateRecentProject
    -- * Recent Projects Array
  , validateRecentProjects
  , safeValidateRecentProjects
  ) where

import GHC.Generics (Generic)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Maybe (Maybe(..))

import Lattice.Schemas.SharedValidation
  ( ValidationError, mkError, maxNameLength
  , validateEntityId, validateNonEmptyString, validateNonNegativeInt
  )

--------------------------------------------------------------------------------
-- Constants
--------------------------------------------------------------------------------

maxRecentProjects :: Int
maxRecentProjects = 100

maxThumbnailSize :: Int
maxThumbnailSize = 10 * 1024 * 1024  -- 10MB

maxTimestamp :: Int
maxTimestamp = 2147483647000  -- Year 2038

--------------------------------------------------------------------------------
-- Recent Project
--------------------------------------------------------------------------------

-- | A recent project entry
data RecentProject = RecentProject
  { rpId :: !Text
  , rpName :: !Text
  , rpThumbnail :: !(Maybe Text)
  , rpModifiedAt :: !Int
  }
  deriving stock (Eq, Show, Generic)

-- | Validate RecentProject
validateRecentProject :: RecentProject -> Either ValidationError RecentProject
validateRecentProject rp = do
  _ <- validateEntityId "id" (rpId rp)
  _ <- validateNonEmptyString "name" maxNameLength (rpName rp)
  case rpThumbnail rp of
    Just thumb | T.length thumb > maxThumbnailSize ->
      Left $ mkError "thumbnail" $
        "must be at most " <> T.pack (show maxThumbnailSize) <> " characters"
    _ -> pure ()
  _ <- validateNonNegativeInt "modifiedAt" maxTimestamp (rpModifiedAt rp)
  Right rp

-- | Safe validation
safeValidateRecentProject :: RecentProject -> Maybe RecentProject
safeValidateRecentProject rp =
  case validateRecentProject rp of
    Right p -> Just p
    Left _ -> Nothing

--------------------------------------------------------------------------------
-- Recent Projects Array
--------------------------------------------------------------------------------

-- | Validate array of recent projects
validateRecentProjects :: [RecentProject] -> Either ValidationError [RecentProject]
validateRecentProjects projects
  | length projects > maxRecentProjects =
      Left $ mkError "recentProjects" $
        "must have at most " <> T.pack (show maxRecentProjects) <> " projects"
  | otherwise = do
      mapM_ validateRecentProject projects
      Right projects

-- | Safe validation
safeValidateRecentProjects :: [RecentProject] -> Maybe [RecentProject]
safeValidateRecentProjects projects =
  case validateRecentProjects projects of
    Right ps -> Just ps
    Left _ -> Nothing
